middlewares/mirror.go

```diff
@@ -0,0 +1,149 @@
+package middlewares
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"sync/atomic"
+
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	"github.com/vulcand/oxy/forward"
+	"github.com/vulcand/oxy/utils"
+	"net/url"
+)
+
+//Mirror is a middleware that provides the request mirroring capabilities
+type Mirror struct {
+	BackendURL     *url.URL
+	RequestHeaders map[string]string
+	SampleRate     uint64
+	mirror         *forward.Forwarder
+	counter        uint64
+}
+
+//NewMirrorMiddleware initializes the Mirror for the request mirroring
+func NewMirrorMiddleware(frontend *types.Frontend, backend *types.Backend) (*Mirror, error) {
+	mirror := frontend.Mirror
+	// Default sample rate is every request.
+	var sampleRate uint64 = 1
+	if mirror.SampleRate > 0 {
+		sampleRate = uint64(mirror.SampleRate)
+	}
+	mirrorURL, err := url.Parse(backend.Servers["mirror"].URL)
+	if err != nil {
+		return nil, err
+	}
+
+	forwarder, err := forward.New(
+	//forward.Stream(true),
+	//forward.PassHostHeader(frontend.PassHostHeader),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Mirror{BackendURL: mirrorURL,
+			RequestHeaders: mirror.RequestHeaders,
+			SampleRate:     sampleRate,
+			mirror:         forwarder},
+		nil
+}
+
+func (m *Mirror) ServeHTTP(w http.ResponseWriter, req *http.Request, next http.HandlerFunc) {
+	if m.counter%m.SampleRate == 0 {
+		atomic.AddUint64(&m.counter, 1)
+		// Copy the request and forward to 'next' and to 'mirror'
+		mReq, err := m.copyRequest(req)
+		if err != nil {
+			log.Errorf("Encountered an error while cloning the request: %v", err)
+		}
+
+		// We don't care about the response from the mirror, so we can just launch it in a safe go routine
+		// In case of failure, will log the error
+		safe.GoWithRecover(func() {
+			mRw := createDummyResponseWriter()
+			m.mirror.ServeHTTP(mRw, mReq)
+		}, func(err interface{}) {
+			log.Errorf("Error in request mirroring routine: %s", err)
+		})
+
+		next.ServeHTTP(w, req)
+	} else {
+		// no mirroring
```

Comment:
We also have to increase the `m.counter` here.

Reply:
Done